### PR TITLE
Refactor conv to share code between reference and optimized kernels

### PR DESCRIPTION
**Pull request by njeffrie**
Summary:

Move shared structs / helper functions into conv_common.cc
Clean up some of the existing code to directly call the reference implementations (made possible by the refactor of the helper functions).

Tested with:
make -j8 -f tensorflow/lite/micro/tools/make/Makefile test_kernel_conv_test

make -j8 -f tensorflow/lite/micro/tools/make/Makefile OPTIMIZED_KERNEL_DIR=cmsis_nn TARGET=stm32f4 test_kernel_conv_test

make -f tensorflow/lite/micro/tools/make/Makefile -j8 TARGET=xtensa OPTIMIZED_KERNEL_DIR=xtensa TARGET_ARCH=hifimini XTENSA_CORE=mini1m1m_RG test_kernel_conv_test

### CODE REVIEW

File: tensorflow/lite/micro/kernels/conv_common.cc [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        reinterpret_cast<int*>(data->per_channel_output_shift),
+        output_channels));
+  }
+  return kTfLiteOk;
+}
+
+void* InitConv(TfLiteContext* context, const char* buffer, size_t length) {
+  TFLITE_DCHECK(context->AllocatePersistentBuffer != nullptr);
+  return context->AllocatePersistentBuffer(context, sizeof(OpDataConv));
+}
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/conv_common.cc:122]: remove? It is not declared in conv.h

File: tensorflow/lite/micro/kernels/xtensa/conv.cc [hunk 2]
```diff
@@ -263,59 +234,21 @@ inline void Conv1x32Input32x32Filter(
...
   TFLITE_DCHECK(node->user_data != nullptr);
   TFLITE_DCHECK(node->builtin_data != nullptr);
   auto* params = reinterpret_cast<TfLiteConvParams*>(node->builtin_data);
 
-  TfLiteTensor* output = GetOutput(context, node, kOutputTensor);
-  const TfLiteTensor* input = GetInput(context, node, kInputTensor);
-  const TfLiteTensor* filter = GetInput(context, node, kFilterTensor);
+  TfLiteTensor* output = GetOutput(context, node, kConvOutputTensor);
+  const TfLiteTensor* input = GetInput(context, node, kConvInputTensor);
+  const TfLiteTensor* filter = GetInput(context, node, kConvWeightsTensor);
```
Comment (advaitjain) [inline tensorflow/lite/micro/kernels/xtensa/conv.cc:249]: This looks identical to the Prepare in the reference kernel.

You do have PrepareConv in conv_common.cc but not in conv.h. Did you mean to share this code between the reference and xtensa kernels?

As a minor nit, at least in softmax.h we are going with the naming convention of SoftmaxInit and SoftmaxPrepare.

**Review by advaitjain — CHANGES_REQUESTED** [review]
some of my comments are for the existing code and not just the refactor -- apologies for that.

File: tensorflow/lite/micro/kernels/xtensa/conv.cc (hunk 2 above)
Comment (njeffrie) [reply tensorflow/lite/micro/kernels/xtensa/conv.cc:249]: I opted not to split out prepare entirely since some implementations require additional logic (for example cmsis_nn). CalculateOpDataConv seems like a decent compromise.

File: tensorflow/lite/micro/kernels/BUILD [hunk 3]
```diff
@@ -136,6 +136,43 @@ cc_library(
     }),
 )
 
+cc_library(
+    name = "conv",
+    srcs = [
```
[the remark this answers is not in the document]
Comment (njeffrie) [reply tensorflow/lite/micro/kernels/BUILD:141]: Done.

File: tensorflow/lite/micro/kernels/conv.cc [hunk 4]
```diff
@@ -178,144 +92,70 @@ TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
   data->output_zero_point = output->params.zero_point;
```
[the remark this answers is not in the document]
Comment (njeffrie) [reply tensorflow/lite/micro/kernels/conv.cc:92]: Done.

File: tensorflow/lite/micro/kernels/conv.h [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  int32_t* per_channel_output_multiplier;
+  int32_t* per_channel_output_shift;
+
+  // The range of the fused activation layer. For example for kNone and
+  // uint8_t these would be 0 and 255.
+  int32_t output_activation_min;
+  int32_t output_activation_max;
+
+  // Index to buffer for optimizations if applicable.
+  int buffer_idx;
```
[the remark this answers is not in the document]
Comment (njeffrie) [reply tensorflow/lite/micro/kernels/conv.h:50]: Done.

File: tensorflow/lite/micro/kernels/conv_common.cc [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+  op_params.stride_width = params->stride_width;
+  op_params.stride_height = params->stride_height;
+  op_params.dilation_width_factor = params->dilation_width_factor;
+  op_params.dilation_height_factor = params->dilation_height_factor;
+  return op_params;
+}
+
+// Returns a ConvParams struct with all the parameters needed for a
+// quantized computation.
+ConvParams ConvParamsQuantized(TfLiteConvParams* params,
```
[the remark this answers is not in the document]
Comment (njeffrie) [reply tensorflow/lite/micro/kernels/conv_common.cc:57]: Done.

File: tensorflow/lite/micro/kernels/kernel_util.h [hunk 7]
```diff
@@ -69,6 +70,18 @@ const RuntimeShape GetTensorShape(const TfLiteEvalTensor* tensor);
...
+  switch (padding) {
+    case TfLitePadding::kTfLitePaddingSame:
+      return PaddingType::kSame;
+    case TfLitePadding::kTfLitePaddingValid:
+      return PaddingType::kValid;
+    case TfLitePadding::kTfLitePaddingUnknown:
+    default:
+      return PaddingType::kNone;
+  }
+}
```
[the remark this answers is not in the document]
Comment (njeffrie) [reply tensorflow/lite/micro/kernels/kernel_util.h:83]: Done.

File: tensorflow/lite/micro/kernels/conv_common.cc [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+  TF_LITE_ENSURE(context, input != nullptr);
+  const TfLiteTensor* filter = GetInput(context, node, kConvWeightsTensor);
+  TF_LITE_ENSURE(context, filter != nullptr);
+
+  int input_width = input->dims->data[2];
+  int input_height = input->dims->data[1];
+  int filter_width = filter->dims->data[2];
+  int filter_height = filter->dims->data[1];
+  int output_width = output->dims->data[2];
+  int output_height = output->dims->data[1];
```
[the remark this answers is not in the document]
Comment (njeffrie) [reply tensorflow/lite/micro/kernels/conv_common.cc:143]: Removed this method.

File: tensorflow/lite/micro/kernels/conv_common.cc [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+                      affine_quantization->zero_point->size);
+  }
+
+  TF_LITE_ENSURE_STATUS(CalculateOpDataConv(
+      context, node, params, input_width, input_height, filter_width,
+      filter_height, output_width, output_height, input->type, data));
+
+  data->input_zero_point = input->params.zero_point;
+  data->filter_zero_point = filter->params.zero_point;
+  data->output_zero_point = output->params.zero_point;
```
[the remark this answers is not in the document]
Comment (njeffrie) [reply tensorflow/lite/micro/kernels/conv_common.cc:179]: Done.

File: tensorflow/lite/micro/kernels/conv_common.cc (hunk 1 above)
Comment (njeffrie) [reply tensorflow/lite/micro/kernels/conv_common.cc:122]: Done.

File: tensorflow/lite/micro/kernels/conv_common.cc [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+  // All per-channel quantized tensors need valid zero point and scale arrays.
+  if (input->type == kTfLiteInt8) {
+    TF_LITE_ENSURE_EQ(context, filter->quantization.type,
+                      kTfLiteAffineQuantization);
+
+    const auto* affine_quantization =
+        static_cast<TfLiteAffineQuantization*>(filter->quantization.params);
+    TF_LITE_ENSURE(context, affine_quantization);
+    TF_LITE_ENSURE(context, affine_quantization->scale);
+    TF_LITE_ENSURE(context, affine_quantization->zero_point);
```
[the remark this answers is not in the document]
Comment (njeffrie) [reply tensorflow/lite/micro/kernels/conv_common.cc:163]: Done.

**Review by advaitjain — APPROVED** [review]
